colossus/src/main/scala/colossus/protocols/http/StreamingHttp.scala

```diff
@@ -0,0 +1,231 @@
+case class End[T <: HttpMessageHead[T]]() extends StreamBodyMessage[T]
+
+
+trait StreamHttp extends Protocol {
+
+
+  type Request = StreamHttpMessage[HttpRequestHead]
+  type Response = StreamHttpMessage[HttpResponseHead]
+}
+
+trait HeadParserProvider[T <: HttpMessageHead[T]] {
+  def parser: Parser[T]
+  def eosTerminatesMessage : Boolean
+}
+object HeadParserProvider {
+  implicit object RequestHeadParserProvider extends HeadParserProvider[HttpRequestHead] {
+    def parser = HttpRequestParser.httpHead
+
+    def eosTerminatesMessage = false
+
+    
+  }
+  object ResponseHeadParserProvider extends HeadParserProvider[HttpResponseHead] {
+    def parser = HttpResponseParser.head
+
+    def eosTerminatesMessage = true
+  }
+}
+
+trait StreamDecoder[T <: HttpMessageHead[T]] { 
+
+  def parserProvider: HeadParserProvider[T]
+
+  private var headParser = parserProvider.parser
+
+  private sealed trait State
+  private case object HeadState extends State
+  private sealed trait BodyState extends State {
+    def nextPiece(input: DataBuffer): Option[StreamBodyMessage[T]]
+  }
+  //used when the transfer-encoding is identity or a content-length is provided
+  //(so non-chunked encoding)
+  private class FiniteBodyState(val size: Option[Int]) extends BodyState {
+    private var taken = 0
+    def done = size.map{_ == taken}.getOrElse(false)
+    def nextPiece(input: DataBuffer): Option[StreamBodyMessage[T]] = if (done) {
+      Some(End())
+    } else if (input.hasUnreadData) {
+      val bytes = input.take(size.map{_ - taken}.getOrElse(input.remaining))
+      taken += bytes.length
+      Some(BodyData(DataBlock(bytes), false))
+    } else {
+      None
+    }
+  }
+  private class ChunkedBodyState extends BodyState {
+    import parsing.Combinators._
+    val parser: Parser[StreamBodyMessage[T]] = intUntil('\r', 16) <~ byte |> {
+      case 0 => bytes(2) >> {_ => End()}
+      case n => bytes(n.toInt) <~ bytes(2) >> {bytes => BodyData(DataBlock(bytes), false)}
+    }
+    def nextPiece(input: DataBuffer): Option[StreamBodyMessage[T]] = parser.parse(input)
+  
+  }
+  private var state: State = HeadState
+
+  def decode(data: DataBuffer): Option[StreamHttpMessage[T]] = state match {
+    case HeadState => headParser.parse(data) match {
+      case Some(h) => {
+        if (h.headers.transferEncoding == TransferEncoding.Chunked) {
+          state = new ChunkedBodyState
+        } else {
+          val lengthOpt = if (parserProvider.eosTerminatesMessage || h.headers.contentLength.isDefined) {
+            h.headers.contentLength
+          } else {
+            Some(0)
+          }
+          state = new FiniteBodyState(lengthOpt)
+        }
+        Some(Head(h))
+      }
+      case None => None
+    }
+    case b: BodyState => b.nextPiece(data) match {
+      case Some(End()) => {
+        state = HeadState
+        Some(End())
+      }
+      case other => other
+    }
+  }
+
+
+  def resetDecoder() {
```

Comment:
For clarity I like to see all methods provide a return type. I'm guessing in this case it is Unit?